### PR TITLE
feat: remote backend

### LINKED ISSUE #88: Add remote backend mode from bifrost-gateway
Filling here so we don't forget, upstream details in https://github.com/ipfs/boxo/issues/576.

Once we have block / car backends in `boxo/gateway`, it will be easy to add config flag to rainbow which disables libp2p stack and only uses delegated retrieval via HTTP (trustless gateway + delegated routing).

I think the basic configuration options would be:
- `RAINBOW_REMOTE_BACKEND` with list of URLs of trustless gateways
- `RAINBOW_REMOTE_BACKEND_MODE=block|car` where `block` is implicit default (if not set)

### PR DESCRIPTION
**Pull request by hacdias**
On top of #111. Closes #88. This PR adds support for remote backends (as described in #88). I made it such that it is the most flexible as I could. It should in scenarios where Libp2p is completely disabled (no bitswap, no dht), but also if libp2p needs to be enabled for things such as peer routing for peering and such.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -289,57 +315,66 @@ share the same seed as long as the indexes are different.
...
+		var peeringAddrs []peer.AddrInfo
+		var index int
 		var err error
 
-		credDir := os.Getenv("CREDENTIALS_DIRECTORY")
-		secretsDir := ddir
+		bitswap := cctx.Bool("bitswap")
+		dhtRouting := DHTRouting(cctx.String("dht-routing"))
+		seedPeering := cctx.Bool("seed-peering")
+		noLibp2p := !bitswap && dhtRouting == DHTOff && !seedPeering
```
Comment (lidel) [inline main.go:325]: :thought_balloon:  `noLibp2p` setup feels overly complicated, and baloons the surface or permutations we need to test. 

In practice either user will run their own libp2p node for dht+bitswap, or they will run no libp2p at all, and delegate both retrieval and routing to remote HTTP endpoints.

@hacdias Could you tweak it to have an explicit, global `RAINBOW_LIBP2P=true|false` (with `true` being default) that controls when `SetupNoLibp2p` is used instead of `Setup[WithLibp2p]`? 

Then, inside `SetupNoLibp2p` just check if `bitswap`, `dht-routing`,  `seed-peering`, `libp2p` flags are set to `true`, and error otherwise.

We want end user's configuration to be simple:
`RAINBOW_LIBP2P=false RAINBOW_REMOTE_BACKENDS=http://trustless-gateway.link ./rainbow`

**Review by lidel — CHANGES_REQUESTED** [review]
Thanks, nearly there :)

We need to streamline configuration a bit –  fetch my changes, I've tweaked some docs, but need your help with  comments inline.

Second ask is to extend gateway-conformance.yml Github Actions workflow  in this repo.

In addition to current test (with lib2p), it should run separate backend tests similar to ones from https://github.com/ipfs/bifrost-gateway/blob/0bb1a904cc4af6c0b5e3398a67f320c1eec383a1/.github/workflows/gateway-conformance.yml#L14

File: main.go (hunk 1 above)
Comment (hacdias) [reply main.go:325]: > Then, inside `SetupNoLibp2p` just check if `bitswap`, `dht-routing`,  `seed-peering`, `libp2p` flags are set to `true`, and error otherwise.

Shouldn't they be `false` and `dht-routing=off`? If we want to check if they haven't been set, we need to do that before calling `SetupNoLibp2p`.

Comment (hacdias) [reply main.go]: I have pushed a commit with a global flag

File: setup_routing.go [hunk 2]
```diff
@@ -197,6 +210,25 @@ func setupRouting(ctx context.Context, cfg Config, h host.Host, ds datastore.Bat
 	return cr, pr, vs, nil
 }
 
+func setupRoutingNoLibp2p(cfg Config, dnsCache *cachedDNS) (routing.ValueStore, error) {
+	delegatedRouters, err := setupDelegatedRouting(cfg, dnsCache)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(cfg.RemoteBackends) > 0 && cfg.RemoteBackendMode == RemoteBackendBlock {
```
[the remark this answers is not in the document]
Comment (hacdias) [reply setup_routing.go:219]: Honestly, no idea why I have the comparison for block. Will update.

File: .github/workflows/gateway-conformance.yml
Comment (lidel) [inline .github/workflows/gateway-conformance.yml:1]: :information_source: refactored so we run separate test for `car` and `block` modes

File: main.go [hunk 3]
```diff
@@ -289,57 +327,77 @@ share the same seed as long as the indexes are different.
...
-			seed = strings.TrimSpace(string(seedBytes))
+		// as a convenience to the end user, and to reduce confusion
+		// libp2p is disabled when remote backends are defined
+		remoteBackends := cctx.StringSlice("remote-backends")
+		if len(remoteBackends) > 0 {
+			fmt.Printf("RAINBOW_REMOTE_BACKENDS set, forcing RAINBOW_LIBP2P=false\n")
+			libp2p = false
+			bitswap = false
+			dhtRouting = DHTOff
 		}
```
Comment (lidel) [inline main.go:348]: :information_source: ended up with this, to ensure we are not running into unexpected edge cases that we missed in test coverage, or may have in future after someone refactors things.

TLDR: for end user, logic is simple: setting a remote backend effectively disables libp2p and related features, and only HTTP backend and routers remain.

End user UX is:

```console
$ RAINBOW_REMOTE_BACKENDS=http://127.0.0.1:8080 ./rainbow
Starting rainbow dev-build
RAINBOW_REMOTE_BACKENDS set, forcing RAINBOW_LIBP2P=false
```

**Review by lidel — APPROVED** [review]
Thank you once again for salvaging this work @hacdias :heart: 

I've pushed some test/docs, and simplified UX a bit, but kept both `LIBP2P` and `BITSWAP` flags, as they will be useful for testing when we start experimenting with HTTP retrieval.

Please take a look, and if no concerns, feel free to merge.

(we want to ship this as v1.3.0, but probably need to land #138  first)

**Comment by lidel** [comment]
Merged https://github.com/ipfs/github-mgmt/pull/205 and removed old conformance as required, @hacdias  you should be able to merge now. If so, mind shipping as v1.3?
